docker file folder creation

cdiop opened the pull request.

cdiop committed:
docker file folder creation

tobwiens commented:
@cdiop The code is okay as a first step. The content of the file is the same or very similar to what the original classes container, variables names are correctly renamed. 

I can start a review as soon as the classes are adjusted to their use.

You need to adjust the copied and renamed classes to do what you want them to do.

cdiop commented:
Retest this please

cdiop committed:
implement script engine

tobwiens commented on src/main/java/jsr223/docker/compose/utils/CommandlineOptionsFromBindingsExtractor.java:
```diff
@@ -53,6 +54,12 @@
 
     public final static String DOCKER_COMPOSE_COMMANDLINE_OPTIONS_SPLIT_REGEX_DEFAULT = " ";
 
+    public final static String DOCKER_FILE_COMMANDLINE_OPTIONS_SPLIT_REGEX_DEFAULT = " ";
```
You can remove `public final static String DOCKER_COMPOSE_UP_COMMANDLINE_OPTIONS_KEY = "docker-compose-up-options";`, `public final static String DOCKER_COMPOSE_COMMANDLINE_OPTIONS_KEY = "docker-compose-options";`, `public final static String DOCKER_COMPOSE_COMMANDLINE_OPTIONS_KEY = "docker-compose-options";`, `public final static String DOCKER_COMPOSE_COMMANDLINE_OPTIONS_SPLIT_REGEX_KEY = "docker-compose-options-split-regex";`

tobwiens replied:
And methods which are connected with it. Or if we need the same functionality with the docker file command then you can keep it but rename it.

cdiop replied:
This file is shared by the two script engines

tobwiens commented on src/main/java/jsr223/docker/compose/utils/CommandlineOptionsFromBindingsExtractor.java:
```diff
@@ -98,4 +105,40 @@
         options.put(OptionType.UP_OPTION, upCmdOptions);
         return options;
     }
+
```
Looks good, I think you can remove the "old" code

cdiop replied:
This file is shared by the two script engines. I keep it in the utils folder

tobwiens commented on src/main/java/jsr223/docker/file/DockerFileScriptEngine.java:
```diff
@@ -0,0 +1,164 @@
+@Log4j
+public class DockerFileScriptEngine extends AbstractScriptEngine {
+
+    private static final String DOCKER_HOST_PROPERTY_NAME = "DOCKER_HOST";
```
Did you have a look what this is in the docker-compose script engine. Not sure if we need it here.

tobwiens commented on src/main/java/jsr223/docker/file/DockerFileScriptEngine.java:
```diff
@@ -0,0 +1,164 @@
+
+    private ProcessBuilderUtilities processBuilderUtilities = new ProcessBuilderUtilities();
+
+    private VariablesReplacer variablesReplacer = new VariablesReplacer();
```
I think we don't need to replace variables

fviale replied:
@tobwiens it's not true, we do need to replace variables

fviale replied:
@cdiop after a discussion with tobias, this can be handled by a generic info containing parameters given to docker

tobwiens commented on src/main/java/jsr223/docker/file/DockerFileScriptEngine.java:
```diff
@@ -0,0 +1,164 @@
+        variablesMap.put(DOCKER_HOST_PROPERTY_NAME, DockerFilePropertyLoader.getInstance().getDockerHost());
+
+        // Replace variables in configuration file
+        String scriptReplacedVariables = variablesReplacer.replaceVariables(script, variablesMap);
```
I think we don't need to replace variables inside the Dockerfile. We should use bash and the --build-arg options to provide variables in a docker fashion.

tobwiens commented on src/main/java/jsr223/docker/file/DockerFileScriptEngine.java:
```diff
@@ -0,0 +1,164 @@
+
+        File dockerfile = null;
+
+        Thread shutdownHook = null;
```
I think you can remove the shutdown hook

tobwiens commented on src/main/java/jsr223/docker/file/DockerFileScriptEngine.java:
```diff
@@ -0,0 +1,164 @@
+                }
+            }
+            if (shutdownHook != null) {
+                Runtime.getRuntime().removeShutdownHook(shutdownHook);
```
I think we don't need  to handle a shutdown hook for the `docker build` command.

fviale replied:
@tobwiens @cdiop 
we need to be able to destroy inside a shutdownhook. the docker Process started from Java. Otherwise this process can become zombie when the task is killed.

tobwiens replied:
Right :+1:

tobwiens replied:
From the docker documentation `If the Docker client loses connection to the daemon, the build is canceled. This happens if you interrupt the Docker client with CTRL-c or if the Docker client is killed for any reason. If the build initiated a pull which is still running at the time the build is cancelled, the pull is cancelled as well.` 

So it's a very good idea to kill/interrupt the client

tobwiens commented on src/main/java/jsr223/docker/file/DockerFileScriptEngine.java:
```diff
@@ -0,0 +1,164 @@
+        Thread shutdownHook = null;
+
+        try {
+            dockerfile = configurationFileWriter.forceFileToDisk(scriptReplacedVariables,
```
Could you have a look if we could provide the Dockerfile to the docker build command without writing it to disk?

tobwiens commented on src/main/java/jsr223/docker/file/DockerFileScriptEngineFactory.java:
```diff
@@ -0,0 +1,132 @@
+ */
+package jsr223.docker.file;
+
+import java.util.*;
```
Please remove/replace the `.*` star import

cdiop replied:
I will have several import if I remove the *

cdiop committed:
docker file logic

cdiop committed:
run built image

cdiop committed:
manage build image, run container and clean image and container

fviale committed:
Allow usage of docker on windows

 - don't use linux docker commands absolute path on windows environments

fviale committed:
Improve DockerFile and DockerCompose script engines

  - improve docker file & compose properties handling: allow overriding by system properties, handle windows & mac docker-compose default path, added a property to keep the docker files
  - improve docker compose script engine: run the docker-compose command in the scratch space (mandatory in non-forked mode).
  - improve docker file script engine: run the docker commands in the scratch space, use job & task ids to tag containers & images, allow to specify command line options through generic information
  - added non-mocked tests for both engines

fviale committed:
Merge pull request #35 from fviale/dockerfile

Allow usage of docker on windows
